<a href="https://colab.research.google.com/github/danielarodriguez4/Stress-Sleep-Patterns-Analysis/blob/main/sp_5_Daniela_Rodr%C3%ADguez_David_Agudelo.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img src="https://github.com/mariabda2/intro_data_2026/blob/main/sesiones_practicas/imgs/banner_fcd.jpg?raw=1" alt="bannersp" width="1100"  height="150">

## <span style="color:#2F749F;"><strong>📦 Sesión práctica 5: Imputación de datos</strong></span>
---
<p align="right">
  <a href="https://colab.research.google.com/github/mariabda2/intro_data_2026/blob/main/sesiones_practicas/sp_5_maria_bernarda_salazar.ipynb?clone=true" target="_blank">
    <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Abrir en Colab"/>
  </a>
</p>

Relacione sus datos y los del compañero de trabajo:

* María Daniela Rodríguez Chacón
mdaniela.rodriguez@udea.edu.co

* David Agudelo Ochoa
david.agudeloo@udea.edu.co

## <span style="color:#2F749F;"><strong>Ejercicio 1. Limpieza e imputación de un conjunto de datos mixto</strong></span>

1. Crea el DataFrame base **df** ejecutando el siguiente código:
    ```python
    df = pd.DataFrame({
        'Nombre': ['Ana', 'Luis', 'Pedro', None, 'Marta', 'Luis', 'Sofía'],
        'Edad': [25, np.nan, 35, 29, -5, 25, None],
        'Ciudad': ['Bogotá', 'Medellín', None, 'Medellín', 'Cali', 'Bogotá', 'Cali'],
        'Ingreso': [3500, 4800, np.nan, 5200, 5100, np.nan, 4700],
        'FechaIngreso': ['2023-01-01', '2023-01-05', None, '2023-01-10', '2023-01-12', None, '2023-01-15']
    })

2. En un nuevo notebook, aplica las etapas del proceso de limpieza e imputación de datos que consideres necesarias sobre el DataFrame df. Ten encuenta incluir:
    - Código correctamente comentado y ejecutado.
    - Explicaciones breves de cada paso.
    - Resultados visuales o estadísticos que evidencien las transformaciones.

In [ ]:
import pandas as pd
import numpy as np

# Creación del DataFrame
df = pd.DataFrame({
    'Nombre': ['Ana', 'Luis', 'Pedro', None, 'Marta', 'Luis', 'Sofía'],
    'Edad': [25, np.nan, 35, 29, -5, 25, None],
    'Ciudad': ['Bogotá', 'Medellín', None, 'Medellín', 'Cali', 'Bogotá', 'Cali'],
    'Ingreso': [3500, 4800, np.nan, 5200, 5100, np.nan, 4700],
    'FechaIngreso': ['2023-01-01', '2023-01-05', None, '2023-01-10', '2023-01-12', None, '2023-01-15']
})

In [ ]:
df

,Nombre,Edad,Ciudad,Ingreso,FechaIngreso
0,Ana,25.0,Bogotá,3500.0,2023-01-01
1,Luis,NaN,Medellín,4800.0,2023-01-05
2,Pedro,35.0,None,NaN,None
3,None,29.0,Medellín,5200.0,2023-01-10
4,Marta,-5.0,Cali,5100.0,2023-01-12
5,Luis,25.0,Bogotá,NaN,None
6,Sofía,NaN,Cali,4700.0,2023-01-15


In [ ]:
print("Valores nulos por columna:")
print(df.isnull().sum())

Valores nulos por columna:
Nombre          1
Edad            2
Ciudad          1
Ingreso         2
FechaIngreso    2
dtype: int64


In [ ]:
# Eliminamos filas duplicadas exactas
df = df.drop_duplicates().reset_index(drop=True)
print(f"Registros después de eliminar duplicados: {len(df)}")

Registros después de eliminar duplicados: 7


In [ ]:
# Convertimos valores negativos en NaN para luego imputarlos correctamente
df.loc[df['Edad'] <= 0, 'Edad'] = np.nan

# Verificamos el cambio
df['Edad']

,Edad
0,25.0
1,NaN
2,35.0
3,29.0
4,NaN
5,25.0
6,NaN


In [ ]:
# Imputación de Edad e Ingreso con la mediana
df['Edad'] = df['Edad'].fillna(df['Edad'].median())
df['Ingreso'] = df['Ingreso'].fillna(df['Ingreso'].median())

# Imputación de Ciudad con la moda (el primer valor de la serie)
df['Ciudad'] = df['Ciudad'].fillna(df['Ciudad'].mode()[0])

# Imputación de Nombre
df['Nombre'] = df['Nombre'].fillna("Desconocido")

print("Estado de nulos después de la imputación:")
df[['Nombre', 'Edad', 'Ciudad', 'Ingreso']].isnull().sum()

Estado de nulos después de la imputación:


,0
Nombre,0
Edad,0
Ciudad,0
Ingreso,0


In [ ]:
# Conversión a datetime
df['FechaIngreso'] = pd.to_datetime(df['FechaIngreso'])

# Imputación de fechas
df['FechaIngreso'] = df['FechaIngreso'].fillna(method='ffill')

# Asegurar que Edad sea un entero
df['Edad'] = df['Edad'].astype(int)

/tmp/ipykernel_11651/1104623782.py:5: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df['FechaIngreso'] = df['FechaIngreso'].fillna(method='ffill')


In [ ]:
print("Dataframe limpio")
df

Dataframe limpio


,Nombre,Edad,Ciudad,Ingreso,FechaIngreso
0,Ana,25,Bogotá,3500.0,2023-01-01
1,Luis,27,Medellín,4800.0,2023-01-05
2,Pedro,35,Bogotá,4800.0,2023-01-05
3,Desconocido,29,Medellín,5200.0,2023-01-10
4,Marta,27,Cali,5100.0,2023-01-12
5,Luis,25,Bogotá,4800.0,2023-01-12
6,Sofía,27,Cali,4700.0,2023-01-15


## <span style="color:#2F749F;"><strong>Ejercicio 2. Limpieza de duplicados</strong></span>

1. Crea el DataFrame base **df** ejecutando el siguiente código:
    ```python
    df = pd.DataFrame({
        'ID': [101, 102, 102, 103, 104, 104, 104],
        'Nombre': ['Ana', 'Luis', 'Luis', 'Marta', 'Carlos', 'Carlos', 'Carlos'],
        'Edad': [25, 30, 30, 29, 40, 40, 41],
        'Ciudad': ['Bogotá', 'Cali', 'Cali', 'Medellín', 'Cali', 'Cali', 'Cali'],
        'FechaRegistro': ['2023-01-01', '2023-01-05', '2023-01-05', '2023-01-10',
                        '2023-01-15', '2023-01-15', '2023-01-16']
    })

2.  En el mismo notebook anterior, para el nuevo dataframe df, responde a las siguientes preguntas (utilizando python):
    - ¿Cuál es el total de registros originales?
    - ¿Cuáles y cuántos son los duplicados exactos?
    - ¿Cuáles y cuántos son los duplicados por varias columnas?
    - ¿Cuántos registros debes eliminar?
    - ¿Cuántos registros quedan después de la limpieza?

In [ ]:
import pandas as pd

# Creación del Dataframe
df = pd.DataFrame({
    'ID': [101, 102, 102, 103, 104, 104, 104],
    'Nombre': ['Ana', 'Luis', 'Luis', 'Marta', 'Carlos', 'Carlos', 'Carlos'],
    'Edad': [25, 30, 30, 29, 40, 40, 41],
    'Ciudad': ['Bogotá', 'Cali', 'Cali', 'Medellín', 'Cali', 'Cali', 'Cali'],
    'FechaRegistro': ['2023-01-01', '2023-01-05', '2023-01-05', '2023-01-10',
                    '2023-01-15', '2023-01-15', '2023-01-16']
})

#### **¿Cuál es el total de registros originales?**

In [ ]:
total_original = len(df)
print(f"Total de registros originales: {total_original}")

Total de registros originales: 7


#### **¿Cuáles y cuántos son los duplicados exactos?**

In [ ]:
# Identificar filas que son copias exactas
duplicados_exactos = df[df.duplicated(keep=False)]
cant_exactos = df.duplicated().sum()

print(f"Cantidad de filas que son duplicados exactos: {cant_exactos}")
print("Registros duplicados exactos:")
print(duplicados_exactos)

Cantidad de filas que son duplicados exactos: 2
Registros duplicados exactos:
    ID  Nombre  Edad Ciudad FechaRegistro
1  102    Luis    30   Cali    2023-01-05
2  102    Luis    30   Cali    2023-01-05
4  104  Carlos    40   Cali    2023-01-15
5  104  Carlos    40   Cali    2023-01-15


#### **¿Cuáles y cuántos son los duplicados por varias columnas?**

In [ ]:
# Buscamos registros donde el ID y Nombre se repitan
duplicados_columnas = df[df.duplicated(subset=['ID', 'Nombre'], keep=False)]
cant_columnas = df.duplicated(subset=['ID', 'Nombre']).sum()

print(f"Cantidad de registros con ID y nombre repetido: {cant_columnas}")
print("Registros con ID/Nombre duplicado:")
print(duplicados_columnas)

Cantidad de registros con ID y nombre repetido: 3
Registros con ID/Nombre duplicado:
    ID  Nombre  Edad Ciudad FechaRegistro
1  102    Luis    30   Cali    2023-01-05
2  102    Luis    30   Cali    2023-01-05
4  104  Carlos    40   Cali    2023-01-15
5  104  Carlos    40   Cali    2023-01-15
6  104  Carlos    41   Cali    2023-01-16


#### **¿Cuántos registros debes eliminar?**

In [ ]:
# Registros a eliminar para que el ID sea único
registros_a_eliminar = total_original - df.drop_duplicates(subset=['ID']).shape[0]
print(f"Registros sugeridos para eliminar: {registros_a_eliminar}")

Registros sugeridos para eliminar: 3


#### **¿Cuántos registros quedan después de la limpieza?**

In [ ]:
df_limpio = df.drop_duplicates(subset=['ID'], keep='last') # Mantenemos el registro más reciente
print(f"Registros finales: {len(df_limpio)}")
print(df_limpio)

Registros finales: 4
    ID  Nombre  Edad    Ciudad FechaRegistro
0  101     Ana    25    Bogotá    2023-01-01
2  102    Luis    30      Cali    2023-01-05
3  103   Marta    29  Medellín    2023-01-10
6  104  Carlos    41      Cali    2023-01-16


## <span style="color:#2F749F;"><strong>Ejercicio 3. Corrección de errores tipográficos o de codificación</strong></span>


1. Crea el DataFrame base **df** ejecutando el siguiente código:
    ```python
    df = pd.DataFrame({
        'Ciudad': ['bogota', 'Bogotá', 'BOGOTA', 'bogotá', 'bogata', 'Bógota', 'BogoTa', 'Cali', 'calí', 'medellín', 'medellin']
    })

2.  En el mismo notebook anterior, para el nuevo dataframe df, estandariza la variable ciudad utilizando python.

In [ ]:
import pandas as pd

# Creación del DataFrame con inconsistencias
df = pd.DataFrame({
    'Ciudad': ['bogota', 'Bogotá', 'BOGOTA', 'bogotá', 'bogata', 'Bógota', 'BogoTa', 'Cali', 'calí', 'medellín', 'medellin']
})

print(f"Categorías únicas iniciales: {df['Ciudad'].nunique()}")
print(df['Ciudad'].unique())

Categorías únicas iniciales: 11
['bogota' 'Bogotá' 'BOGOTA' 'bogotá' 'bogata' 'Bógota' 'BogoTa' 'Cali'
 'calí' 'medellín' 'medellin']


**Estrategia de Estandarización**

Para resolver esto, seguiremos estos pasos:

1.  **Normalización de mayúsculas:** todo a minúsculas o formato título.
2.   **Eliminación de espacios:**   quitar espacios en blanco al inicio o final (strip).
1.   **Eliminación de tildes:** usar una función para quitar acentos.
2.   **Corrección de errores tipográficos:** reemplazar manualmente valores erróneos (como "bogata").












In [ ]:
import unicodedata

# Función para eliminar tildes y caracteres especiales
def normalizar_texto(texto):
    if pd.isna(texto):
        return texto
    # Normalizar a forma NFD (separa la letra del acento) y filtrar los caracteres de combinación
    texto = unicodedata.normalize('NFD', texto)
    texto = texto.encode('ascii', 'ignore').decode("utf-8")
    return texto.lower().strip()

# Aplicamos la normalización básica
df['Ciudad_Limpia'] = df['Ciudad'].apply(normalizar_texto)

# Corrección de errores específicos (typos)
# Reemplazamos 'bogata' por la forma correcta
df['Ciudad_Limpia'] = df['Ciudad_Limpia'].replace({'bogata': 'bogota'})

# Finalmente, pasamos a Capitalize para estética (Bogota, Cali, Medellin)
df['Ciudad_Final'] = df['Ciudad_Limpia'].str.capitalize()

print("\nCategorías únicas finales:")
print(df['Ciudad_Final'].unique())


Categorías únicas finales:
['Bogota' 'Cali' 'Medellin']


## <span style="color:#2F749F;"><strong>Ejercicio 4. Identificación de mecanismos de datos faltantes (MCAR, MAR, MNAR)</strong></span>
1. Se tiene una base de datos de estudiantes universitarios con las siguientes variables: , Edad, Género, Promedio académico, Ingreso familiar y Horas de estudio por semana. Sin embargo, hay valores faltantes en algunas variables.
2. En los siguientes escenarios, clasifica el tipo de dato faltante (MCAR, MAR o MNAR) y justifica tu respuesta en máximo 2 líneas.
   - **Escenario 1.**  Al analizar los datos, se observa que los valores faltantes en Ingreso familiar aparecen de forma completamente aleatoria, sin relación con ninguna otra variable (ni edad, ni género, ni promedio, etc.).
   - **Escenario 2.** Se detecta que los valores faltantes en Horas de estudio ocurren principalmente en estudiantes con promedios bajos.
   - **Escenario 3.** Se detecta que los valores faltantes en Horas de estudio ocurren principalmente en estudiantes con promedios bajos.
   - **Escenario 4.** Se observa que los valores faltantes en Promedio académico aparecen con mayor frecuencia en estudiantes de cierto género.
   - **Escenario 5.** Algunos estudiantes con muy pocas horas de estudio no reportan esta variable porque consideran que el valor es “demasiado bajo” para registrarlo.
   

*   **Escenario 1**

**Clasificación:** MCAR (Missing Completely At Random). La falta de datos no depende de ninguna variable observada ni del valor de la variable misma; es puramente fruto del azar.

*   **Escenario 2 y 3**

**Clasificación:** MAR (Missing At Random). La probabilidad de que falte el dato depende de otra variable observada en el sistema (el promedio académico), no del azar.

*   **Escenario 4**

**Clasificación:** MAR (Missing At Random). El patrón de ausencia está sistemáticamente relacionado con una característica conocida del estudiante (el género).

*  **Escenario 5**


**Clasificación:** MNAR (Missing Not At Random). El dato falta precisamente por el valor que debería haber sido registrado; existe una decisión consciente o sesgo basado en el dato oculto.

## <span style="color:#2F749F;"><strong>Ejercicio 5. Imputación con KNN y regresión</strong></span>

1. Crea el DataFrame base ejecutando el siguiente código:

    ```python
        import pandas as pd
        import numpy as np

        np.random.seed(42)

        n = 30

        df = pd.DataFrame({
            'Edad': np.random.randint(18, 60, n),
            'Promedio': np.round(np.random.uniform(2.0, 5.0, n), 2),
            'Horas_estudio': np.random.randint(0, 40, n),
            'Ingreso_familiar': np.random.randint(800000, 8000000, n),
            'Semestres_cursados': np.random.randint(1, 10, n)
        })

        # Introducción de valores faltantes con distintos patrones
        df.loc[[2, 5, 8, 15], 'Promedio'] = np.nan
        df.loc[[1, 7, 10, 18], 'Horas_estudio'] = np.nan
        df.loc[[3, 6, 14], 'Ingreso_familiar'] = np.nan

2. Al analizar el dataframe `df`:
   - ¿Qué variables tienen mayor proporción de datos faltantes?
   - ¿Observas algún patrón estructurado o parecen aleatorios?

3. Realiza la imputación de valores faltantes utilizando KNN y regresión
4. Analiza si se preserva la naturaleza de los datos con cada una de las técnicas de imputación (KNN y regresión)

In [ ]:
# Análisis de datos faltantes
import pandas as pd
import numpy as np
from sklearn.impute import KNNImputer
from sklearn.linear_model import LinearRegression

# Recreación del DataFrame
np.random.seed(42)
n = 30
df = pd.DataFrame({
    'Edad': np.random.randint(18, 60, n),
    'Promedio': np.round(np.random.uniform(2.0, 5.0, n), 2),
    'Horas_estudio': np.random.randint(0, 40, n),
    'Ingreso_familiar': np.random.randint(800000, 8000000, n),
    'Semestres_cursados': np.random.randint(1, 10, n)
})

# Introducción de nulos
df.loc[[2, 5, 8, 15], 'Promedio'] = np.nan
df.loc[[1, 7, 10, 18], 'Horas_estudio'] = np.nan
df.loc[[3, 6, 14], 'Ingreso_familiar'] = np.nan

# Análisis de faltantes
print(df.isnull().mean() * 100)

Edad                   0.000000
Promedio              13.333333
Horas_estudio         13.333333
Ingreso_familiar      10.000000
Semestres_cursados     0.000000
dtype: float64


**¿Qué variables tienen mayor proporción de datos faltantes?**

Las variables con mayor proporción de nulos son `Promedio` y `Horas_estudio`, ambas con un 13.33% (correspondiente a 4 registros de los 30 totales). Le sigue `Ingreso_familiar` con un 10% (3 registros). Las variables `Edad` y `Semestres_cursados` están completas (0%).



In [ ]:
df

,Edad,Promedio,Horas_estudio,Ingreso_familiar,Semestres_cursados
0,56,2.70,35.0,2909818.0,7
1,46,2.27,NaN,3031785.0,7
2,32,NaN,30.0,5283302.0,9
3,25,3.15,14.0,NaN,3
4,38,4.95,7.0,3525928.0,7
5,56,NaN,13.0,4653339.0,1
6,36,4.58,22.0,NaN,4
7,40,4.04,NaN,1879111.0,4
8,28,NaN,20.0,1796107.0,5
9,28,2.04,15.0,4321441.0,7


**¿Observas algún patrón estructurado o parecen aleatorios?**

Los datos parecen seguir un patrón aleatorios. No se observa que la falta de un dato dependa del valor de otra variable (por ejemplo, los nulos en "Ingreso_familiar" ocurren tanto en estudiantes jóvenes como mayores). Es una distribución dispersa diseñada para probar la robustez de los algoritmos de imputación.

In [ ]:
# Imputación de valores faltantes usando KNN y regresión
# Copia del DF para KNN
df_knn = df.copy()

# Inicializamos el imputador (usaremos 3 vecinos para un dataset pequeño)
imputer_knn = KNNImputer(n_neighbors=3)
df_knn_imputed = pd.DataFrame(imputer_knn.fit_transform(df_knn), columns=df_knn.columns)

print("Muestra de datos imputados con KNN (Fila 2 - Promedio):")
print(df_knn_imputed.loc[2, 'Promedio'])

# -------- Verificación --------
print()
print("Valores faltantes antes de imputar:")
print(df_knn.isnull().sum())

print("\nValores faltantes después de imputar:")
print(df_knn_imputed.isnull().sum())

Muestra de datos imputados con KNN (Fila 2 - Promedio):
3.473333333333333

Valores faltantes antes de imputar:
Edad                  0
Promedio              4
Horas_estudio         4
Ingreso_familiar      3
Semestres_cursados    0
dtype: int64

Valores faltantes después de imputar:
Edad                  0
Promedio              0
Horas_estudio         0
Ingreso_familiar      0
Semestres_cursados    0
dtype: int64


In [ ]:
# Imputación por regreison
df_reg = df.copy()

def imputar_regresion(dataset, columna_objetivo):
    # Separamos filas con y sin nulos
    test_data = dataset[dataset[columna_objetivo].isnull()]
    train_data = dataset.dropna()

    # Columnas predictoras (todas menos la que queremos imputar)
    X_train = train_data.drop(columns=[columna_objetivo])
    y_train = train_data[columna_objetivo]
    X_test = test_data.drop(columns=[columna_objetivo])

    X_test = X_test.fillna(X_train.mean())

    modelo = LinearRegression()
    modelo.fit(X_train, y_train)
    predicciones = modelo.predict(X_test)

    dataset.loc[dataset[columna_objetivo].isnull(), columna_objetivo] = predicciones

# Imputamos las 3 columnas con nulos
for col in ['Promedio', 'Horas_estudio', 'Ingreso_familiar']:
    imputar_regresion(df_reg, col)

print("Muestra de datos imputados con Regresión (Fila 2 - Promedio):")
print(df_reg.loc[2, 'Promedio'])

# Verificación
print()
print("Valores faltantes antes de imputar:")
print(df.isnull().sum())

print("\nValores faltantes después de imputar:")
print(df_reg.isnull().sum())

Muestra de datos imputados con Regresión (Fila 2 - Promedio):
3.2581080732047574

Valores faltantes antes de imputar:
Edad                  0
Promedio              4
Horas_estudio         4
Ingreso_familiar      3
Semestres_cursados    0
dtype: int64

Valores faltantes después de imputar:
Edad                  0
Promedio              0
Horas_estudio         0
Ingreso_familiar      0
Semestres_cursados    0
dtype: int64


In [ ]:
df_reg

,Edad,Promedio,Horas_estudio,Ingreso_familiar,Semestres_cursados
0,56,2.700000,35.000000,2.909818e+06,7
1,46,2.270000,20.157851,3.031785e+06,7
2,32,3.258108,30.000000,5.283302e+06,9
3,25,3.150000,14.000000,4.409997e+06,3
4,38,4.950000,7.000000,3.525928e+06,7
5,56,3.030432,13.000000,4.653339e+06,1
6,36,4.580000,22.000000,4.205366e+06,4
7,40,4.040000,21.983912,1.879111e+06,4
8,28,3.454357,20.000000,1.796107e+06,5
9,28,2.040000,15.000000,4.321441e+06,7


**Analiza si se preserva la naturaleza de los datos con cada una de las técnicas de imputación (KNN y regresión)**

Para determinar si se preserva la naturaleza de los datos, debemos observar lo siguiente:

**1. Con KNN:**

Es excelente para mantener la consistencia local. Al usar vecinos, el valor imputado (3.47) está dentro del rango real de los datos y se basa en perfiles similares.


**2. Con regresión:**

Preserva la relación estadística (correlación) entre variables. Si en la vida real el promedio sube cuando suben las horas de estudio, la regresión mantendrá esa lógica de manera estricta.

## <span style="color:#2F749F;"><strong>Ejercicio 6. Imputación con LOCF y NOCB</strong></span>

1. Crea el DataFrame base ejecutando el siguiente código:

    ```python
        import pandas as pd
        import numpy as np

        np.random.seed(42)

        n = 30

        df = pd.DataFrame({
            'Fecha': pd.date_range(start='2023-01-01', periods=n, freq='D'),
            'Ventas': np.round(np.random.uniform(100, 500, n), 2),
            'Inventario': np.random.randint(10, 100, n),
            'Clientes': np.random.randint(1, 50, n)
        })

        # Introducción de valores faltantes
        df.loc[[3, 4, 5, 12, 13], 'Ventas'] = np.nan
        df.loc[[7, 8, 20], 'Inventario'] = np.nan
        df.loc[[1, 2, 18, 19], 'Clientes'] = np.nan
        df = df.sort_values('Fecha').reset_index(drop=True)

2. Al analizar el DataFrame `df`:
   - ¿Qué variables presentan datos faltantes consecutivos?
   - ¿Por qué este tipo de patrón es relevante en series temporales?
   - ¿Qué implicaciones tiene el orden de los datos para este tipo de imputación?
3. Realiza la imputación de valores faltantes utilizando LOCF y NOCB
4. ¿Cuál método consideras más apropiado para este caso y por qué?

In [ ]:
import pandas as pd
import numpy as np

np.random.seed(42)

n = 30

df = pd.DataFrame({
    'Fecha': pd.date_range(start='2023-01-01', periods=n, freq='D'),
    'Ventas': np.round(np.random.uniform(100, 500, n), 2),
    'Inventario': np.random.randint(10, 100, n),
    'Clientes': np.random.randint(1, 50, n)
})

# Introducción de valores faltantes
df.loc[[3, 4, 5, 12, 13], 'Ventas'] = np.nan
df.loc[[7, 8, 20], 'Inventario'] = np.nan
df.loc[[1, 2, 18, 19], 'Clientes'] = np.nan
df = df.sort_values('Fecha').reset_index(drop=True)

In [ ]:
df.head()

,Fecha,Ventas,Inventario,Clientes
0,2023-01-01,249.82,60.0,29.0
1,2023-01-02,480.29,16.0,NaN
2,2023-01-03,392.80,30.0,NaN
3,2023-01-04,NaN,82.0,44.0
4,2023-01-05,NaN,48.0,34.0


In [ ]:
df_nan = df[df.isna().any(axis=1)]
print(df_nan)

        Fecha  Ventas  Inventario  Clientes
1  2023-01-02  480.29        16.0       NaN
2  2023-01-03  392.80        30.0       NaN
3  2023-01-04     NaN        82.0      44.0
4  2023-01-05     NaN        48.0      34.0
5  2023-01-06     NaN        27.0      10.0
7  2023-01-08  446.47         NaN      14.0
8  2023-01-09  340.45         NaN      31.0
12 2023-01-13     NaN        62.0      14.0
13 2023-01-14     NaN        11.0      23.0
18 2023-01-19  272.78        17.0       NaN
19 2023-01-20  216.49        56.0       NaN
20 2023-01-21  344.74         NaN      24.0


**¿Qué variables presentan datos faltantes consecutivos?**

Como se puede observa en el `df_nan`, las variables `Ventas`, `Inventario` y `Clientes` presentan datos faltantes consecutivos.

* `Ventas`: 2 rachas de datos faltantes
* `Inventario`: 1 racha de datos faltantes
* `Clientes`: 2 rachas de datos faltantes

**¿Por qué este tipo de patrón es relevante en series temporales?**

Porque en una serie temporal, los datos tienen un orden, por lo que hay que tener en cuenta no solo cuántos faltan, sino también cuándo y durante cuánto tiempo.

**¿Qué implicaciones tiene el orden de los datos para este tipo de imputación?**

El orden de los datos es fundamental porque en una serie temporal existe una dependencia cronológica; el valor de hoy suele estar relacionado con el de ayer. Si se altera el orden o se ignora, se pierde la estructura de la serie y se pueden generar datos imputados que no tienen sentido lógico con la tendencia o la estacionalidad del fenómeno.